main/src/library/Nec.flix

```diff
@@ -0,0 +1,909 @@
+///
+pub enum Nec[t] {
+    case NecOne(t)
+    case Nec(Nec[t], Nec[t])
+}
+
+instance Eq[Nec[a]] with Eq[a] {
+    pub def eq(c1: Nec[a], c2: Nec[a]): Bool = Nec.equals(c1, c2)
+}
+
+instance Hash[Nec[a]] with Hash[a] {
+    pub def hash(c: Nec[a]): Int32 = 39119 + Hash.hash(Nec.toList(c))
+}
+
+instance SemiGroup[Nec[a]] {
+    pub def combine(c1: Nec[a], c2: Nec[a]): Nec[a] = Nec.append(c1, c2)
+}
+
+instance Functor[Nec] {
+    pub def map(f: a -> b & ef, c: Nec[a]): Nec[b] & ef = Nec.map(f, c)
+}
+
+instance Applicative[Nec] {
+    pub def point(x: a) : Nec[a] = Nec.singleton(x)
+    pub def ap(f: Nec[a -> b & ef], x: Nec[a]) : Nec[b] & ef = Nec.ap(f, x)
+}
+
+instance Monad[Nec] {
+    pub def flatMap(f: a -> Nec[b] & ef, x : Nec[a]) : Nec[b] & ef = Nec.flatMap(f, x)
+}
+
+instance Foldable[Nec] {
+    pub def foldLeft(f: (b, a) -> b & ef, s: b, c: Nec[a]): b & ef = Nec.foldLeft(f, s, c)
+    pub def foldRight(f: (a, b) -> b & ef, s: b, c: Nec[a]): b & ef = Nec.foldRight(f, s, c)
+}
+
+instance Traversable[Nec] {
+    pub def traverse(f: a -> m[b] & ef, t: Nec[a]): m[Nec[b]] & ef with Applicative[m] = Nec.traverse(f, t)
+    pub override def sequence(t: Nec[m[a]]): m[Nec[a]] with Applicative[m] = Nec.sequence(t)
+}
+
+instance Reducible[Nec] {
+    pub def reduceLeftTo(f: (b, a) -> b & ef1, g: a -> b & ef2, l: Nec[a]): b & (ef1 and ef2) = Nec.reduceLeftTo(f, g, l)
+    pub def reduceRightTo(f: (a, b) -> b & ef1, g: a -> b & ef2, l: Nec[a]): b & (ef1 and ef2) = Nec.reduceRightTo(f, g, l)
+}
+
+instance Iterable[Nec] {
+    pub def iterator(t: Nec[a]): Iterator[a] & Impure =
+        Nec.toIterator(t)
+}
+
+instance ToString[Nec[a]] with ToString[a] {
+    pub def toString(c: Nec[a]): String =
+        let fn = () -> {
+            let sb = StringBuilder.new();
+            StringBuilder.appendString!(sb, "Nec#{");
+            Nec.foldLeft((acc, x) -> {
+                if (acc < 1)
+                    StringBuilder.appendString!(sb, "${x}")
+                else
+                    StringBuilder.appendString!(sb, ", ${x}");
+                acc + 1}, 0, c);
+            StringBuilder.appendString!(sb, "}");
+            StringBuilder.toString(sb)
+        };
+        fn() as & Pure
+}
+
+
+namespace Nec {
+
+    use Functor.{<$>};
+    use Applicative.{<*>};
+
+    ///
+    /// A datatype for pattern matching on a Nec (traversing left-to-right).
+    ///
+    pub enum ViewLeft[a] with Eq {
+        case OneLeft(a)
+        case SomeLeft(a, Nec[a])
+    }
+
+    ///
+    /// A datatype for pattern matching on a Nec (traversing right-to-left).
+    ///
+    pub enum ViewRight[a] with Eq {
+        case OneRight(a)
+        case SomeRight(Nec[a], a)
+    }
+
+    ///
+    /// Returns `true` if and only if `c1` and `c2` and equal.
+    ///
+    pub def equals(c1: Nec[a], c2: Nec[a]): Bool with Eq[a] =
+        def loop(nec1, nec2) =
+            match (viewLeft(nec1), viewLeft(nec2)) {
```

Comment:
Can you just use the `Eq` instance on `ViewLeft`?

Reply:
I think `ViewLeft` (and `ViewRight`) need to use the `Eq` instance on `Nec`. 

Basically the shape of the `Nec` should be opaque and `Nec` s can be equal if they contain the same elements when viewed from either the left of the right. The internal binary tree might have a different shape according to how the tree was built - this is why we have a hand coded instance for `Nec` and don't derive `Eq`.

Reply:
I guess I'm a little confused.

I understand that we can't derive `Eq` for `Nec`. But can't we do the following?

```scala
    pub def equals(c1: Nec[a], c2: Nec[a]): Bool with Eq[a] = viewLeft(nec1) == viewLeft(nec2)
```

I believe that this should do the same thing as what you have written, as the derived `Eq[ViewLeft[a]]` will call the hand-coded `Eq[Nec[a]]` (which then calls the derived `Eq[ViewLeft[a]]`...).

Reply:
Thanks Matthew, yes - that works so I've updated the equals definition and added some equals and not-equals tests.